### PR TITLE
CI: upload artifacts

### PR DESCRIPTION
**Pull request by umarcor**
In this PR the artifacts generated in CI jobs are uploaded using `actions/upload-artifact@v2`. This feature was already implemented in `windows_builds.yml`, but it was commented out.

For each push or pull request, all successfully generated artifacts will be uploaded in the run. This will allow users to try new features and/or fixes without needing to build Godot themselves.

This is a partial implementation of godotengine/godot-proposals#1412. See also godotengine/godot-proposals#1458.

### CODE REVIEW

**Comment by HaSa1002** [comment]
The problem was and is the limited GitHub Actions Storage. Thats why it is commented out.

**Comment by umarcor** [comment]
@HaSa1002 was there any actual issue before having it commented or was it commented preventively? As discussed in https://github.com/godotengine/godot-proposals/issues/1412#issuecomment-680173450, I believe there is no limitation. Can you please provide any specific reference?

**Comment by Calinou** [comment]
> Can you please provide any specific reference?

GitHub Actiosn storage is limited to 1 GB in the free tier.

I wonder if we could use https://github.com/c-hive/gha-remove-artifacts on each run to remove all artifacts except the latest one… However, this will prevent it using for PR builds.

**Comment by Calinou** [comment]
After doing some testing, it seems **GitHub Actions storage is not limited for public repositories**. Here's some proof:

I set up a repository with the following workflow file:

```yaml
name: 📊Test
on: [push, pull_request]

jobs:
  static-checks:
    name: Generate random binary file and upload to artifacts
    runs-on: ubuntu-20.04
    steps:
      - name: Checkout
        uses: actions/checkout@v2

      - name: Generate random binary file
        run: |
          dd if=/dev/urandom of=100M.dat bs=1048576 count=100

      - name: Upload artifacts
        uses: actions/upload-artifact@v2
        with:
          name: download
          path: |
            100M.dat
            README.md
```

I made 10+ commits using the following command:

``bash
export COUNT=10 && rm -f README.md && echo "Artifact $COUNT" > README.md && git commit -am "Artifact $COUNT" && git push
```

I got the following builds, all with successful artifact uploads:

![image](https://user-images.githubusercontent.com/180032/94681471-cc39a280-0323-11eb-81cf-7ad0ad0c967b.png)

My Billing page still says I'm not using any artifact storage:

![image](https://user-images.githubusercontent.com/180032/94681499-db205500-0323-11eb-8132-f1963bdfc55f.png)

**Comment by umarcor** [comment]
@Calinou, see [About billing for GitHub Actions](https://docs.github.com/en/free-pro-team@latest/github/setting-up-and-managing-billing-and-payments-on-github/about-billing-for-github-actions):

> **GitHub Actions usage is free for public repositories and self-hosted runners**. For private repositories, each GitHub account receives a certain amount of free minutes and storage, depending on the product used with the account.

This proposal is to be used in public repositories only. Hence, if I understand correctlly, it is free, without explicit limits.

> Also, GitHub might still not like you if you have 20+ GB of artifacts in a single repository 🙂

I don't see why they would have an issue with that. 20GB is a rather small figure. Users have reported uploading artifacts of 1-5GB (each day). 

> It would be interesting to check if this limit is enforced in public repositories

My fork, which I used to open this PR, is a public repository for testing. Shall I add a CRON job every hour?

**Comment by Calinou** [comment]
@umarcor I've just done testing, see my comment above. This PR is probably good to go :slightly_smiling_face:

**Review by Calinou — APPROVED** [review]

**Comment by Calinou** [comment]
As for the "nightly builds" part, isn't this provided by the `master` branch being built on every commit?

**Comment by umarcor** [comment]
> @umarcor I've just done testing, see my comment above. 

I'm sorry, I didn't see it, even after posting after you. Thanks for testing :D

> To give an example, GitHub has a soft limit of 1 GB per repository (for Git objects, that is). You can go above this limit, but you might get an email from GitHub if you go way overboard. This is especially true if your repository has low activity and doesn't appear to be used by other people. (Thankfully, this isn't the case of Godot 🙂)
> 
> I've never gotten such an email, but I know it's happened to other people in the past.

Good to know. My point is that I did some not good usage of GitHub Actions when I was learning. E.g., running 30 workflows at the same time or triggering a lot of chained workflows (an infinite loop). In both cases, the service was temporaly disabled, as if I had reached an API rate-limit, and it was ok again in a few hours. Since then, I believe they changed the approach. Now they silently ignore and/or remove older content instead. Therefore, I think that the problem we might have in this repo is too many PRs fighting to be the last updated. Hopefully, we won't have more than 20 PRs fighting for the same half-day timeframe.

Nonetheless, we might limit this to non PRs. I believe the main benefit is for users to get the latest master. Should anyone want to try a PR, all they need to do is pushing the branch to their fork, and they'll have artifacts built there.

> This PR is probably good to go.

I'd like someone with more experience to check if the naming of the artifacts makes sense. Furthermore, `bin/*` is uploaded in all cases. I don't know if that's correct.

There is an specific work which behaves strange. In Android builds, `bin/*` is reported to be empty: https://github.com/godotengine/godot/actions/runs/279681997. However, restarting the job seems to fix it: https://github.com/umarcor/godot/actions/runs/279559788

> As for the "nightly builds" part of [godotengine/godot-proposals#1412](https://github.com/godotengine/godot-proposals/issues/1412), isn't this provided by the `master` branch being built on every commit? Either way, I think that proposal should be closed in favor of [godotengine/godot-proposals#1458](https://github.com/godotengine/godot-proposals/issues/1458) if this PR is merged.

The difference between just building master on every commit or a "nightly release" is the consistency of the URL. GitHub Actions does not provide an URL for "retrieving an specific artifact from the latest successful workflow run". Therefore, users need to browse the latest runs for finding the artifacts. However, artifacts of successful runs of branch master can be uploaded as assets of an specific release (nightly). Then, https://github.com/godotengine/godot/releases/download/nightly/ASSET_NAME.EXT would always point to the latest artifacts. See, for example, https://github.com/ghdl/ghdl/releases/tag/nightly. That would allow users to test their projects against latest Godot in CI, without building Godot themselves.

Anyway, I agree that godotengine/godot-proposals#1412 can be closed in favour of godotengine/godot-proposals#1458 when/if this PR is merged.

**Comment by Calinou** [comment]
> Nonetheless, we might limit this to non PRs. I believe the main benefit is for users to get the latest master. Should anyone want to try a PR, all they need to do is pushing the branch to their fork, and they'll have artifacts built there.

This is a good point. It might be difficult for beginners to find the exact link, but pull request authors can edit their initial post to link to the builds more directly.

**Comment by umarcor** [comment]
> This is a good point. It might be difficult for beginners to find the exact link, but pull request authors can edit their initial post to link to the builds more directly.

So, shall we keep this PR on hold until I do the following?

- I modify it for artifacts to be uploaded on non-PR events only.
- I add some documentation (would be glad to be suggested where) explaining contributors and users how to get builds for specific branches/PRs using CI in their forks.

**Comment by Calinou** [comment]
@umarcor It's now possible to shorten the artifact retention period: https://github.blog/changelog/2020-10-08-github-actions-ability-to-change-retention-days-for-artifacts-and-logs/

**Comment by umarcor** [comment]
@Calinou, that's really nice! Thanks for letting me know. Shall I investigate how to set different retention periods for master, non-master and PRs? E.g.:

- master: 1 week.
- non-master: 3 days.
- PR: 1 day.

**Comment by Calinou** [comment]
@umarcor Since GitHub currently doesn't limit the storage limit for public repositories, we can be more generous on retention dates. I think we could set the retention limit to 30 days for everything at first and see what happens. If we get a complaint from GitHub, we can then reduce those durations.

**Comment by umarcor** [comment]
@Calinou, done!

**Comment by Xrayez** [comment]
> Since GitHub currently doesn't limit the storage limit for public repositories, we can be more generous on retention periods. I think we could set the retention limit to 14 days for everything at first and see what happens. If we get a complaint from GitHub, we can then reduce those durations.

I don't understand this logic. If GitHub is capable of storing those artifacts without limits for free, it means that there should be no problem with it and that should be by design. If we get a complaint, I say that GitHub is pretentious in this regard.

**Comment by Calinou** [comment]
@Xrayez GitHub (and most SaaS services really) tend to apply "soft limits" on many things like repository size. This means there is no hard limit, but you might get an email from them if one of your repositories grows really large, especially if it has few external contributors (which hints at you using GitHub as a backup service).

File: .github/workflows/android_builds.yml [hunk 1]
```diff
@@ -71,3 +71,9 @@ jobs:
           SCONS_CACHE: ${{github.workspace}}/.scons_cache/
         run: |
           scons -j2 verbose=yes warnings=all werror=yes platform=android target=release tools=no
+
+      - uses: actions/upload-artifact@v2
+        with:
+          name: android-template
```
Comment (Xrayez) [inline .github/workflows/android_builds.yml:77]:
```suggestion
          name: ${{github.job}}
```

Seems to work in the same way. The suggestion applies to other jobs. 🙂

Comment (umarcor) [reply .github/workflows/android_builds.yml]: Done!

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**